### PR TITLE
[PK-210] Splitting ScoreDiscovery into two variants

### PR DESCRIPTION
**Pull request by packpaul**
DefaultScoreDiscovery made full score path aware;
score path concept removed from Score;
some clean-up in unit tests

## Overview

Please describe your changes here.

---

### Checklist

- [x] Change is covered by automated tests.
- [x] JavaDoc is provided for all new/changed public APIs.
- [ ] User Guide is updated.
- [x] CI builds pass.

### CODE REVIEW

**Comment by packpaul** [comment]
Functional testing was performed for score discovery over classpath resources. It is still unclear how to automate this check.

File: celesta-sql/src/main/java/ru/curs/celesta/score/Grain.java [hunk 1]
```diff
@@ -22,12 +21,15 @@
...
 import ru.curs.celesta.CelestaException;
 import ru.curs.celesta.DBType;
+import ru.curs.celesta.score.io.Resource;
 
 /**
  * Grain.
  */
 public final class Grain extends NamedElement {
 
+    private static final Pattern NATIVE_SQL = Pattern.compile("--\\{\\{(.*)--}}", Pattern.DOTALL);
```
Comment (inponomarev) [inline celesta-sql/src/main/java/ru/curs/celesta/score/Grain.java:31]: This is suboptimal regexp for native sql as "greedy dot" can be catastrophically slow (see [this talk](https://www.youtube.com/watch?v=pLKrKFy1t1Y) when you have time).

To write the optimal regexp for native sql was a non-trivial task, the right solution (in JavaCC syntax) is here: https://github.com/CourseOrchestra/celesta/blob/413485c2517f1d9357307f5b8ab1f7aa868babaa/celesta-sql/src/main/javacc/CelestaParser.jj#L142

(and this is [how I derived it](https://twitter.com/inponomarev/status/956663963814645761), if you're curious).

The only aim of this regexp is to strip `--{{` `--}}` from the code. Maybe `startswith`, `endswith` and `substring` will do better.

However, this is not your code, you just  moved it and maybe this whole comment is not relevent to your PR :-)

File: celesta-sql/src/main/java/ru/curs/celesta/score/NamedElement.java [hunk 2]
```diff
@@ -83,7 +84,17 @@ public final int hashCode() {
...
     @Override
     public final boolean equals(Object obj) {
-        return obj instanceof NamedElement ? name.equals(((NamedElement) obj).getName()) : name.equals(obj);
+        if (obj == this) {
+            return true;
+        }
+        if (obj instanceof NamedElement) {
+            return Objects.equals(this.name, ((NamedElement) obj).getName());
+        }
+        if (obj instanceof String) {
```
Comment (inponomarev) [inline celesta-sql/src/main/java/ru/curs/celesta/score/NamedElement.java]: I think the `if (obj instanceof String)` clause can be  safely removed, if no tests fail.
This violates `equals` contract for no reason and SpotBugs is absolutely right nagging about it.

File: celesta-sql/src/main/java/ru/curs/celesta/score/io/UrlResource.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public String toString() {
+        return urlString;
+    }
+
+    @Override
+    public int hashCode() {
+        return urlString.hashCode();
```
Comment (inponomarev) [inline celesta-sql/src/main/java/ru/curs/celesta/score/io/UrlResource.java:57]: 👍

**Review by inponomarev — CHANGES_REQUESTED** [review]

File: celesta-sql/src/test/java/ru/curs/celesta/score/io/ResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Test
+    public void testGetRelativePath() {
+        Resource parent = new FileResource(new File("/parent/"));
+        Resource child = new FileResource(new File("/parent/child/file.test"));
+        Resource nonChild = new FileResource(new File("/parent/child/../../file.test"));
+        
+        String relativePath = parent.getRelativePath(child);
+        assertEquals("child/file.test", relativePath);
```
Comment (inponomarev) [inline celesta-sql/src/test/java/ru/curs/celesta/score/io/ResourceTest.java]: This one fails if run on Windows. Should be 
```java
        assertEquals("child"
                + File.separator
                + "file.test", relativePath);
```

**Review by inponomarev — CHANGES_REQUESTED** [review]

File: celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                try {
+                    scoreFilesInputStream = scoreFilesResource.getInputStream();
+                } catch (IOException ex) {
+                    System.out.println("score index file is missing: " + scoreFilesResource.toString());
+                    continue;
+                }
+
+                try (BufferedReader reader = new BufferedReader(
+                        new InputStreamReader(scoreFilesInputStream, StandardCharsets.UTF_8))) {
+                    Iterable<String> li = reader.lines()::iterator;
```
Comment (inponomarev) [inline celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java]: This looks fancy. Never knew it can be done this way 😃 
But how about more conservative way to read lines from `BufferedReader`?

```java
String gp;
while ((gp = reader.readLine()) != null) {
```

File: celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+        } catch (IOException ex) {
+            throw new CelestaException("Unable to load score files from resources!", ex);
+        }
+
+        return new LinkedHashSet<>(grainNameToResourceMap.values());
+    }
+
+
+    private String getGrainName(String grainPath) {
```
Comment (inponomarev) [inline celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java]: Should it be package-private and unit-tested?

File: celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                        if (existingGrainResource != null) {
+                            throw new CelestaException("Duplicate resources encountered for the grain '%s': %s, %s",
+                                                       grainName, existingGrainResource, grainResource);
+                        }
+                    }
+                }
+            }
+
+        } catch (IOException ex) {
+            throw new CelestaException("Unable to load score files from resources!", ex);
```
Comment (inponomarev) [inline celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java]: It's a real pity, but let's stay calm in this situation and remove the exclamation mark 😄

**Review by inponomarev — CHANGES_REQUESTED** [review]

File: celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    private static final String SCORE_FILES_LOCATION = "score/score.files";
+
+    @Override
+    public Set<Resource> discoverScore() {
+
+        Map<String, Resource> grainNameToResourceMap = new LinkedHashMap<>();
+
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            Enumeration<URL> urls = (classLoader != null)
```
Comment (inponomarev) [inline celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java:31]: Well, I think I have an idea how to test it. 

The only thing that we have to mock here is `Enumeration<URL> urls`. You may either initialize this variable in constructor and then change its value in unit test, or make an overridable getter method and override it in test.

Just invent any set of urls, and then we may check for corner cases like missing score index files and duplicate resources.

File: celesta-sql/src/main/java/ru/curs/celesta/score/NamedElement.java (hunk 2 above)
Comment (packpaul) [reply celesta-sql/src/main/java/ru/curs/celesta/score/NamedElement.java]: Yes, let's try.

File: celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java (hunk 5 above)
Comment (packpaul) [reply celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java]: Yes indeed slipped out of my sight.

File: celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java (hunk 6 above)
Comment (packpaul) [reply celesta-sql/src/main/java/ru/curs/celesta/score/discovery/ScoreByScoreResourceDiscovery.java]: A unit-test added.

**Review by inponomarev — APPROVED** [review]
👍
